File: java/Lista2/ArabRzym.java
//Przekształca liczby 
// z rzymskiego na arabski -- 
// z arabskiego na rzymski -- if integer try catch exeption - else then rzymski is true
// try catch exeptions: ujemna, nie-integer (nie naturalna)

public class ArabRzym
{

    private static final String[] liczby = { "I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M" };
    private static final int[] wartosci = { 1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000 };

    public static int rzym2arab(String rzym) throws ArabRzymException
    {
        int wynik = 0;
        int i = liczby.length - 1;
        int index = 0;

        while (index < rzym.length() && i >= 0)
        {
            if (rzym.startsWith(liczby[i], index))
            {
                wynik += wartosci[i];
                index += liczby[i].length();
            } 
            else
            {
                i--;
            }
        }

        if (arab2rzym(wynik).equals(rzym))
        {
            return wynik;
        }
        else
        {
            throw new ArabRzymException(rzym + " ma niepoprawny format liczby rzymskiej");
        }
    }

    public static String arab2rzym(int arab) throws ArabRzymException
    {
        if (arab < 1 || arab > 3999)
        {
            throw new ArabRzymException("Liczba poza zakresem (1-3999)");
        }

        StringBuilder wynik = new StringBuilder();
        int i = liczby.length - 1;

        while (arab > 0)
        {
            if (arab >= wartosci[i])
            {
                wynik.append(liczby[i]);
                arab -= wartosci[i];
            }
            else
            {
                i--;
            }
        }

        return wynik.toString();
    }
}

class ArabRzymException extends Exception
{
    public ArabRzymException(String message)
    {
        super(message);
    }
}

File: cpp/Lista2/ArabRzym.cpp
#include "ArabRzym.hpp"
#include <string>

const std::vector<std::string> ArabRzym::liczby = {"I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M"};
const std::vector<int> ArabRzym::wartosci = {1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000};

int ArabRzym::rzym2arab(const std::string& rzym)
{
    int wynik = 0;
    size_t i = liczby.size() - 1;
    size_t index = 0;

    while (index < rzym.length() && i < liczby.size())
    {
        if (rzym.substr(index, liczby[i].length()) == liczby[i])
        {
            wynik += wartosci[i];
            index += liczby[i].length();
        } 
        else
        {
            i--;
        }
    }

    if (arab2rzym(wynik) == rzym)
    {
        return wynik;
    }
    else
    {
        throw std::invalid_argument(" ma niepoprawny format liczby rzymskiej");
    }
}

std::string ArabRzym::arab2rzym(int arab)
{
    if (arab < 1 || arab > 3999)
    {
        throw std::out_of_range(" poza zakresem (1-3999)");
    }

    std::string wynik;
    size_t i = liczby.size() - 1;

    while (arab > 0)
    {
        if (arab >= wartosci[i])
        {
            wynik += liczby[i];
            arab -= wartosci[i];
        }
        else
        {
            i--;
        }
    }

    return wynik;
}